### ISSUE TITLE
[MNG-6223] mvn -f outputs invalid error when specifying POM directory

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7643]
**[Charles Gould](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cgould)** opened **[MNG-6223](https://issues.apache.org/jira/browse/MNG-6223?redirect=false)** and commented

Basic multi-module project: common, client, and server modules.

```
$ mvn -f client package
POM file client specified with the -f/--file command line argument does not exist
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building alternate-pom-client 1.0
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ alternate-pom-client ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /Users/cgould/dev/personal/alternate-pom/client/src/main/resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ alternate-pom-client ---
[INFO] Nothing to compile - all classes are up to date
[INFO]
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ alternate-pom-client ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /Users/cgould/dev/personal/alternate-pom/client/src/test/resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ alternate-pom-client ---
[INFO] No sources to compile
[INFO]
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ alternate-pom-client ---
[INFO] No tests to run.
[INFO]
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ alternate-pom-client ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 0.764 s
[INFO] Finished at: 2017-04-29T14:03:22-04:00
[INFO] Final Memory: 11M/309M
[INFO] ------------------------------------------------------------------------
```

There is an error message, but the client module proceeds to build nonetheless.

I don't see the error message when I run:

```
$ mvn -f client/pom.xml package
```

I've verified it doesn't happen on 3.3.9, but it does happen on 3.5.0 and 3.5.0-alpha-1.

Sample project attached.

---

**Affects:** 3.5.0-alpha-1, 3.5.0

**Attachments:**
- [alternate-pom.zip](https://issues.apache.org/jira/secure/attachment/12865677/alternate-pom.zip) (_6.73 kB_)

**Issue Links:**
- [MNG-6191](https://issues.apache.org/jira/browse/MNG-6191) mvn -f complains about illegal readlink option under macOS

- [MNG-6198](https://issues.apache.org/jira/browse/MNG-6198) mvn script fails to locate .mvn directory when pom location specified with -f

- [MNG-5889](https://issues.apache.org/jira/browse/MNG-5889) .mvn directory should be picked when using --file

- [MNG-5338](https://issues.apache.org/jira/browse/MNG-5338) Accept a directory with -f/--file
 (_**"breaks"**_)

0 votes, 5 watchers

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

nice finding: yes, looks like the shell script does not check when argument is a directory instead of a file

there is MNG-5889 core IT to improve to add this case, in addition to fixing the script itself, so this bug does never gets reintroduced: do you want to try to do the fix?

**Comment by jira-importer** [comment]
**[Charles Gould](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cgould)** commented

[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy) Yesterday I grepped [Maven Core](https://git-wip-us.apache.org/repos/asf/maven.git) for the warning message, but I couldn't find it. Was I looking in the wrong repo?

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

the message is not in java code but in `mvn` shell script, in `apache-maven/src/bin`
it is related to MNG-5889: that's why the IT is in https://git-wip-us.apache.org/repos/asf?p=maven-integration-testing.git, in `core-it-suite/src/test/java/org/apache/maven/it/MavenITmng5889FindBasedir.java` and `core-it-suite/src/test/resources/mng-5889-find.mvn/`

**Comment by jira-importer** [comment]
**[Charles Gould](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cgould)** commented

I've since looked at `mvn` and identified the issue there. It's using

```
if [ -f "${arg}" ]
```

instead of

```
if [ -a "${arg}" ]
```

But `mvn.cmd` is a bit different (lots of gotos).

Just to be clear -- if I'm to submit a patch, it should cover `mvn`, `mvn.cmd`, and the integration tests as well, correct?

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

ideally, it should cover everything, yes: the IT to have some tests (and check if it is affected by the issue or not), then the fix(es)
if you are not able to provide everything, doing what you can is better than relying only on others :)

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

fixed both on Windows and Unix
IT in http://git-wip-us.apache.org/repos/asf/maven-integration-testing/commit/981a8104

**Comment by jira-importer** [comment]
**[Jesse Glick](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jglick@netbeans.org)** commented

Was a regression of MNG-5338.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

thank you for the reference: now we did a core IT, we won't have regression on this in the future

**Comment by jira-importer** [comment]
**[Charles Gould](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cgould)** commented

Thanks for getting this fix into 3.5.2!
